http/src/main/scala/org/http4s/blaze/http/http2/Http2FrameSerializer.scala

```diff
@@ -0,0 +1,293 @@
+    buffer.flip()
+
+    buffer
+  }
+
+  def mkRstStreamFrame(streamId: Int, errorCode: Long): ByteBuffer = {
+    require(0 < streamId, "Invalid RST_STREAM stream id")
+    require(0 <= errorCode && errorCode <= Masks.INT32, s"Invalid error code")
+
+    val payloadSize = 4
+    val buffer = BufferTools.allocate(HeaderSize + payloadSize)
+    writeFrameHeader(payloadSize, FrameTypes.RST_STREAM, 0x0, streamId, buffer)
+    buffer.putInt((errorCode & Masks.INT32).toInt)
+    buffer.flip()
+
+    buffer
+  }
+
+  def mkSettingsAckFrame(): ByteBuffer = mkSettingsFrameImpl(true, Seq.empty)
+
+  def mkSettingsFrame(settings: Seq[Setting]): ByteBuffer =
+    mkSettingsFrameImpl(false, settings)
+
+  private[this] def mkSettingsFrameImpl(ack: Boolean, settings: Seq[Setting]): ByteBuffer = {
+    require(!ack || settings.isEmpty, "Setting acknowledgement must be empty")
+
+    val payloadSize = settings.length * 6
+    val buffer = BufferTools.allocate(HeaderSize + payloadSize)
+    val flags = if (ack) Flags.ACK.toInt else 0x0
+
+    writeFrameHeader(payloadSize, FrameTypes.SETTINGS, flags.toByte, 0, buffer)
+    settings.foreach { case Setting(k,v) => buffer.putShort(k.toShort).putInt(v.toInt) }
+
+    buffer.flip()
+    buffer
+  }
+
+  def mkPushPromiseFrame(
+    streamId: Int,
+    promiseId: Int,
+    endHeaders: Boolean,
+    padding: Int,
+    headerBuffer: ByteBuffer
+  ): Seq[ByteBuffer] = {
+    require(streamId != 0x0, "Invalid Stream id for PUSH_PROMISE frame")
+    require(promiseId != 0x0 && promiseId % 2 == 0, "Invalid Stream id for PUSH_PROMISE frame")
+    require(0 <= padding && padding <= 256, "Invalid padding of HEADER frame")
+
+    val padded = 0 < padding
+    var flags = 0x0
+
+    if (endHeaders) {
+      flags |= Flags.END_HEADERS
+    }
+
+    if (padded) {
+      flags |= Flags.PADDED
+    }
+
+
+    // Need4 for the promised stream id, and maybe 1 for the padding size
+    val bufferSize = HeaderSize + 4 + (if (padded) 1 else 0)
+    val buffer = BufferTools.allocate(bufferSize)
+    writeFrameHeader(4 + padding + headerBuffer.remaining,
+      FrameTypes.PUSH_PROMISE, flags.toByte, streamId, buffer)
+
+    if (padded) {
+      // padding of 1 is represented by the padding field and no trailing padding
+      buffer.put((padding - 1).toByte)
+    }
+
+    buffer
+      .putInt(promiseId)
+      .flip()
+
+    buffer::headerBuffer::tailPadding(padding - 1)
+  }
+
+  def mkPingFrame(ack: Boolean, data: Array[Byte]): ByteBuffer = {
+    val PingSize = 8
+    require(data.length == PingSize, "Ping data must be 8 bytes long")
+
+    val flags: Byte = if (ack) Flags.ACK else 0x0
+    val buffer = ByteBuffer.allocate(HeaderSize + PingSize)
+    writeFrameHeader(PingSize, FrameTypes.PING, flags, 0, buffer)
+    buffer
+      .put(data)
+      .flip()
+
+    buffer
+  }
+
+  def mkGoAwayFrame(lastStreamId: Int, error: Http2Exception): Seq[ByteBuffer] = {
+    val msgString = error.getMessage.getBytes(StandardCharsets.UTF_8)
```

Comment:
Are these always UTF-8?  Most of the time the standards refer to US-ASCII, although it's probably going to be the same, and this is a newer spec.  This is probably right, but just asking.

Reply:
The "additional debug data" of the GOAWAY frame doesn't have a prescribed format.